examples/fungible_token/src/lib.rs

```diff
@@ -0,0 +1,309 @@
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+    fn burn(&mut self, account: &ProgramId, amount: u128) {
+        let zero = ProgramId(H256::zero().to_fixed_bytes());
+        if account == &zero {
+            panic!("FungibleToken: Burn from zero address.");
+        }
+        unsafe {
+            self.decrease_total_supply(amount);
+            let old_balance = FUNGIBLE_TOKEN.get_balance(account);
+            // TODO: safe math for underflow detection.
+            self.set_balance(account, old_balance - amount);
+        }
+        let transfer_data = TransferData {
+            from: H256::from_slice(account.as_slice()),
+            to: H256::zero(),
+            amount,
+        };
+        msg::reply(
+            &Event::Transfer(transfer_data),
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+    fn transfer(&mut self, sender: &ProgramId, recipient: &ProgramId, amount: u128) {
+        let zero = ProgramId(H256::zero().to_fixed_bytes());
+        if sender == &zero {
+            panic!("FungibleToken: Transfer from zero address.");
+        }
+        if recipient == &zero {
+            panic!("FungibleToken: Transfer to zero address.");
+        }
+        let sender_balance = self.get_balance(sender);
+        if amount > sender_balance {
+            panic!("FungibleToken: Transfer amount exceeds balance.");
+        }
+        self.set_balance(sender, sender_balance - amount);
+        let recipient_balance = self.get_balance(recipient);
+        self.set_balance(recipient, recipient_balance + amount);
+        let transfer_data = TransferData {
+            from: H256::from_slice(sender.as_slice()),
+            to: H256::from_slice(recipient.as_slice()),
+            amount,
+        };
+        msg::reply(
+            &Event::Transfer(transfer_data),
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+    fn approve(&mut self, owner: &ProgramId, spender: &ProgramId, amount: u128) {
+        let zero = ProgramId(H256::zero().to_fixed_bytes());
+        if owner == &zero {
+            panic!("FungibleToken: Approve from zero address.");
+        }
+        if spender == &zero {
+            panic!("FungibleToken: Approve to zero address.");
+        }
+
+        self.allownces
+            .entry(*owner)
+            .or_default()
+            .insert(*spender, amount);
+        let approve_data = ApproveData {
+            owner: H256::from_slice(owner.as_slice()),
+            spender: H256::from_slice(spender.as_slice()),
+            amount,
+        };
+        msg::reply(
+            &Event::Approval(approve_data),
+            exec::gas_available() - GAS_RESERVE,
+            0,
+        );
+    }
+    fn get_allowance(&self, owner: &ProgramId, spender: &ProgramId) -> u128 {
+        self.allownces
+            .get(owner)
+            .map_or(0u128, |m| *m.get(spender).unwrap_or(&0u128))
```

Comment:
Make it more linear.
```suggestion
        *self.allownces
            .get(owner)
            .and_then(|m| m.get(spender))
            .unwrap_or(&0)
```